src/lib/Libtpp/tpp_router.c

```diff
@@ -1123,196 +1079,179 @@ router_pkt_presend_handler(int tfd, tpp_packet_t *pkt, void *extra)
  *
  */
 static int
-router_pkt_handler(int tfd, void *data, int len, void *c, void *extra)
+router_pkt_handler(int tfd, void *buf, int len, void *c, void *extra)
 {
 	tpp_context_t *ctx = (tpp_context_t *) c;
+	tpp_data_pkt_hdr_t *dhdr = buf;
 	enum TPP_MSG_TYPES type;
 	tpp_chunk_t chunks[2];
 	tpp_router_t *target_router = NULL;
 	int target_fd = -1;
 	tpp_addr_t connected_host;
+	conn_auth_t *authdata = (conn_auth_t *)extra;
 	tpp_addr_t *addr = tpp_get_connected_host(tfd);
-	void *data_out = NULL;
-	size_t len_out = 0;
+	char msg[TPP_GEN_BUF_SZ];
+	short rc;
 
 	if (!addr)
 		return -1;
 
 	memcpy(&connected_host, addr, sizeof(tpp_addr_t));
 	free(addr);
 
-	type = *((unsigned char *) data);
+again:
+	type = dhdr->type;
+	errno = 0;
+
 	if(type >= TPP_LAST_MSG)
 		return -1;
 
-	if (type == TPP_ENCRYPTED_DATA) {
-		char *msgbuf = NULL;
-		conn_auth_t *authdata = (conn_auth_t *)extra;
-
-		if (authdata == NULL) {
-			snprintf(tpp_get_logbuf(), TPP_LOGBUF_SZ, "tfd=%d, No auth data found in connection %s", tfd, tpp_netaddr(&connected_host));
-			tpp_log_func(LOG_CRIT, __func__, tpp_get_logbuf());
-			return -1;
-		}
+	switch (type) {
+		case TPP_ENCRYPTED_DATA: {
+			int len_out;
+			int sz = sizeof(tpp_encrypt_hdr_t);
+			if (authdata == NULL) {
+				tpp_log(LOG_CRIT, __func__, "tfd=%d, No auth data found", tfd);
+				return -1;
+			}
 
-		if (authdata->encryptdef == NULL) {
-			tpp_log_func(LOG_CRIT, __func__, "Connetion doesn't support decryption of data");
-			return -1;
-		}
+			if (authdata->encryptdef == NULL) {
+				tpp_log(LOG_CRIT, __func__, "connetion doesn't support decryption of data");
+				return -1;
+			}
 
-		if (authdata->encryptdef->decrypt_data(authdata->encryptctx, (void *)((char *)data + 1), (size_t)len - 1, &data_out, &len_out) != 0) {
-			return -1;
-		}
+			if (authdata->encryptdef->decrypt_data(authdata->encryptctx, (void *)((char *)buf + sz), (size_t)len - sz, (void *) &dhdr, (size_t *) &len_out) != 0) {
```

Comment:
I think we are passing &dhdr and dhdr is itself a pointer. This means the decrypt_data is supposed to "allocate" the buffer and set the pointer to the NEW buffer and return the length in len_out.

Indeed from your own documentation: Cleanup: Returned data_out should be freed using free() when no more required, as it will be allocated heap memory. So, we are missing the free - need to chat